### PR TITLE
Galloping search for block_iter_seek

### PR DESCRIPTION
**Pull request by kelvinatorr**
Add galloping search algorithm to block_iter_seek. The logic was taken from [this Wikipedia article](https://en.wikipedia.org/wiki/Exponential_search). However unlike the Wikipedia article, this implementation only gallops if the current block iterator's restart_index is not zero which means a seek was already done on this block previously. I made this change after discussing my benchmarks of the algorithm's performance with @cmikk. This change should improve the chance that we gallop when it is most efficient to do so, when the next key to seek to is close by the previous seek. Before this PR, the entire restart array is binary searched on every seek.

I also added additional unit tests for block_iter_seek to ensure it still works as expected.

### CODE REVIEW

File: mtbl/block.c [hunk 1]
```diff
@@ -249,34 +249,88 @@ block_iter_seek_to_last(struct block_iter *bi)
...
+				 * of left
+				 */
+				if (left * 2 >= right) {
+					/* 
+					 * But doubling left moves us past the end of restart array
+					 * so we stop galloping and just binary search from here
+					 */
+					gallop = false;
+				} else {
+					/* Double the value of left and gallop again */
```
Comment (djw1149) [inline mtbl/block.c:308]: Change to "gallop again"

File: mtbl/block.c [hunk 2]
```diff
@@ -249,34 +249,88 @@ block_iter_seek_to_last(struct block_iter *bi)
...
-			 */
+			if (gallop) {
+				/* 
+				 * We galloped too far. The key we are looking for is now
+				 * before the current value of left. So we reset left back to
+				 * its value before the last doubling. But check if we
+				 * doubled at least once. If we did not double even once,
+				 * just set left to 0.
+				 */
+				left = (mid >= (bi->restart_index * 2) ? mid/2 : 0);
```
Comment (djw1149) [inline mtbl/block.c:330]: I would rather see a if/else then a complicated trinary operator

File: t/test-iter-seek.c [hunk 3]
```diff
@@ -68,15 +73,12 @@ int main(int argc, char ** argv) {
 	struct mtbl_iter *merger_iter = mtbl_source_iter(merger_source);
 	assert(merger_iter != NULL);
 
-	test_iter(merger_iter);
+	if(test_iter(merger_iter) != 0) {
```
Comment (djw1149) [inline t/test-iter-seek.c:76]: add space after "if"

File: t/test-iter-seek.c [hunk 4]
```diff
@@ -96,12 +100,190 @@ int main(int argc, char ** argv) {
...
+
+	struct mtbl_iter *iter = mtbl_source_iter(source);
+
+	int32_t last_number = num_keys - 1;
+
+	/* Loop forward from 0 and seek i + jump keys */
+	for (int32_t i = 0; i < num_keys; i++) {
+		if ((i + jump) < last_number) {
+			int32_t target_number = i + jump;
+			if(seek_and_check(iter, i, target_number) != 0) {
```
Comment (djw1149) [inline t/test-iter-seek.c:124]: add space after "if"

**Review by djw1149 — COMMENTED** [review]
Minor changes requested

File: mtbl/block.c (hunk 1 above)
Comment (kelvinatorr) [reply mtbl/block.c:308]: Changed

File: mtbl/block.c (hunk 2 above)
Comment (kelvinatorr) [reply mtbl/block.c:330]: Ok I changed it to a simpler if/else

File: t/test-iter-seek.c (hunk 3 above)
Comment (kelvinatorr) [reply t/test-iter-seek.c:76]: Fixed

**Review by djw1149 — APPROVED** [review]

**Comment by djw1149** [comment]
Note that this PR should be to a new "next" branch, not straight to master IMO.

**Review by cmikk — COMMENTED** [review]
I would suggest de-tangling the galloping search logic into separate code, so the code reads as:
```
left = 0; right = num_restarts;
if (gallop_ready) {
      /* galloping search to set left, right based on restart_index */
} 
/* binary search with left, right */
```
I believe the above will be much easer to follow than weaving the `gallop` state into the binary search loop. There are enough differences between the gallop process and the binary search process that its worthwhile separating them in the code, even at the cost of some repeated `decode_entry` and `bytes_compare` stanzas.

**Comment by kelvinatorr** [comment]
Ok; I separated out the galloping from the binary search. I put decoding an entry from the restart array and comparing it to the target in new separate function to avoid the duplication.

Also I made test3 better.

**Pull request merged**